# Домашнє завдання до теми "Цикл for"

0. Розігрів. Напишіть цикл, який виводить на екран числа від 1 до 10 включно.

In [1]:
for numb in range(11):
  print(numb, end=' ')

0 1 2 3 4 5 6 7 8 9 10 

**Початок покладено!**

1. Задайте список `numbers`, який містить значення від 0 до 9 включно. Використовуючи цикл `for`, обчисліть суму елементів в створеному списку та виведіть фінальну суму на екран.

In [2]:
# Create the list
numbers = list(range(10))
numbers

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [3]:
# Create a for loop to calculate a final sum of elements by use the in-place addition assignment operator:
total = 0
for num in numbers:
    total += num
print(total)

45


2. Використовуючи цикл  `for` виведіть для кожного елементу списку `numbers`,який ми створили в завданні 1, чи цей елемент дорівнює 5. Тобто після виконання циклу на екрані має бути 10 bool значень.

In [5]:
# Method 1
for num in numbers:
    print(num == 5)

False
False
False
False
False
True
False
False
False
False


In [6]:
# Method 2
for num in numbers:
  if num == 5:
    current_value = True
  else:
    current_value = False
  print(num, current_value)

0 False
1 False
2 False
3 False
4 False
5 True
6 False
7 False
8 False
9 False


3. Використовуючи цикл  `for` для кожного елементу списку `numbers`, який ми створили в завданні 1, виявіть, чи цей елемент ділиться на 2 без остачі, та запишіть результат у новий список `numbers_divisible_by_2`.

Увага! Виконання завдання відрізняється від попереднього тим, що тепер ми результат обчислення на кожній ітерації записуємо у новий список. До того ж ми тут виконуємо іншу математичну операцію.

Виведіть фільнальний список `numbers_divisible_by_2` на екран.

In [7]:
# Method 1
numbers_divisible_by_2 = list()

for num in numbers:
  if num % 2 == 0:
    numbers_divisible_by_2.append(True)
  else:
    numbers_divisible_by_2.append(False)
print(numbers_divisible_by_2)

# Method 2
numbers_divisible_by_2 = list()

for num in numbers:
     numbers_divisible_by_2.append(num % 2 == 0)
print(numbers_divisible_by_2)

[True, False, True, False, True, False, True, False, True, False]
[True, False, True, False, True, False, True, False, True, False]


Очікуваний результат: `[True, False, True, False, True, False, True, False, True, False]`

4. Задано два списки в клітинці нижче - `a` і `b`. Створіть цикл, результатом якого буде список `result`, що містить елементи типу bool і який фактично каже,
"чи на цьому індексі елементи наших двох списків співпадають". Приклад:
```
a = [1, 2, 3, 6]
b = [1, 7, 3, 6]
# ми виконали поелементне порівняння двох списків і отримали новий список result
# result містить наступний результат -> [True, False, True, True]
```
В кінці виведіть вміст result на екран.

<details>
<summary>Підказка</summary>

Аби порівняти елементи в циклі - треба створити цикл по індексам елементів. Зважаючи, що наші списки - одної довжини, цей підхід дозволить нам отримати чітко те, що треба.

**Просунутий варіант.** Альтернативно можна створити цикл, який на кожній ітерації має доступ до двох елементів - по одному з кожного списка, для цього треба скористатись функцією `zip` - приклад [тут](https://realpython.com/python-zip-function/#traversing-lists-in-parallel)
</details>


In [9]:
a = [64, 93, 60, 95, 66, 9, 15, 55, 77, 28, 74, 100, 13, 1, 85, 48, 66, 65, 59, 32]
b = [64, 93, 60, 95, 66, 9, 15, 56, 77, 28, 74, 100, 13, 1, 85, 48, 67, 65, 59, 32]

In [10]:
# Method 1
result = []
for num in range(len(a)):
      result.append(a[num] == b[num])
print(result)

# Method 2: with zip() function
result_2 = [x == y for x, y in zip(a, b)]
print(result_2)

[True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, False, True, True, True]
[True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, False, True, True, True]


5. Нижче задано список `numbers`. Виконайте клітинку з його заданням.

  Давайте знайдемо середнє арифметичне значення елементів списку `numbers` округлене до 2 цифр після коми та запишемо його у змінну `numbers_mean`. Для цього виконаємо два простих кроки.

  1) Знайдіть суму всіх елементів списку `numbers` та запишіть її у змінну `numbers_sum`. Можна з використанням циклу, а можна і без.

  2) Обчисліть `numbers_mean` шляхом ділення суми елементів у списку на їх кількість. Кількість елементів в списку - це його довжина.

  Виведіть `numbers_mean`, округлений до 2 цифр після коми, на екран.

  **Підказка**: округлити число можна з використанням вбудованої функції Python [round()](https://docs.python.org/3/library/functions.html#round).

In [11]:
numbers = [42, 51, 29, 2, 13, 44, 80, 66, 16]

In [12]:
# Method 1
numbers_sum = 0
for num in numbers:
  numbers_sum += num
  numbers_mean = numbers_sum / len(numbers)
print(round(numbers_mean, 2))

# Method 2
numbers_sum = sum(numbers)
numbers_mean = round(numbers_sum / len(numbers), 2)
print(numbers_mean)

# Method 3: by Numpy
import numpy as np

mean = np.mean(numbers)
print(round(mean, 2))

38.11
38.11
38.11


Очікуваний результат: `38.11`

6. Задано список `a` і елемент `element`. Перевірте програмно чи є в списку `a` елемент `element` з допомогою інструкцій `if ... else`. Виведіть на екран відповідь "Так, є", якщо цей елемент в списку присутній, або "Ні, нема", якщо його там немає.
Важливо це зробити без використання імпортів будь-яких бібліотек.

  Тут цикла немає :) Але це завдання - підготовче до наступного.

In [14]:
a = [64, 93, 60, 95, 66, 9, 15, 55, 77, 28, 74, 100, 13, 1, 85, 48, 66, 65, 59, 32]
element = 33

In [15]:
if element in a:
  print("Yes, there is")
else:
  print("No, there is not")

No, there is not


7. Тепер у нас не один елемент, а кілька. Вони наведені в змінній `elements`. За допомогою циклу і інструкцій `if ... else` для кожного елемента списку `elements` перевірте, чи він є в списку `a`.
Виведіть на екран відповідь "Так, є елемент <вставити значення елемента>", якщо елемент в списку присутній, або "Ні, нема елемента <вставити значення елемента>", якщо його там немає.

In [16]:
elements = [44, 66, 77, 100]

In [17]:
for el in elements:
  if el in a:
    print(f"Yes, there is element {el} in a !")
  else:
    print(f"No, there is not element {el} in a.")

No, there is not element 44 in a.
Yes, there is element 66 in a !
Yes, there is element 77 in a !
Yes, there is element 100 in a !
